runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -5376,3 +5376,21 @@ J9::CodeGenerator::isMonitorValueType(TR::Node* monNode)
 
    return TR_no;
    }
+
+TR_YesNoMaybe
+J9::CodeGenerator::isMonitorValueBased(TR::Node* monNode)
+   {
+   if (_monitorMapping.find(monNode->getGlobalIndex()) == _monitorMapping.end())
+      return TR_maybe;
+
+   TR_OpaqueClassBlock *clazz = _monitorMapping[monNode->getGlobalIndex()];
+
+   //java.lang.Object class is only set when monitor is java.lang.Object but not its subclass
+   if (clazz == self()->comp()->getObjectClassPointer())
+      return TR_no;
+
```

Comment:
>What if the class is an interface or abstract class that a value-based class could implement or be derived from, respectively? 

That's a very good question. I experimented with the two cases. `isMonitorValueBased()` could return either `TR_maybe` or `TR_no`. 
- `TR_maybe`: because  the object class is not in `_monitorMapping`.
- `TR_no` because it reaches the end of the `isMonitorValueBased()` and VM `J9_IS_J9CLASS_VALUEBASED` returns `false`; not because `if (clazz == self()->comp()->getObjectClassPointer())` is true. 

If the base class is an interface or abstract class, it won't hit the path that `if (clazz == self()->comp()->getObjectClassPointer())` is true. 

pseudo code: 
```
interface A 

@ValueBased
public class B implements A 

Object obj = new B(); 
synchronized(obj) { ... }

A obj = new B(); 
synchronized(obj) { ... }
```
```
abstract class C

@ValueBased
public class D extends C {

Object obj = new D(); 
synchronized(obj) { ... }

C obj = new D(); 
synchronized(obj) { ... }
```

Reply:
Right - these are the kinds of case I was worried about.  I was thinking there should be an additional check of `isConcreteClass`, as is done in [J9::CodeGenerator::isMonitorValueType](https://github.com/eclipse/openj9/blob/3b44a9e833681f5914bc4d10690fbe7107393329/runtime/compiler/codegen/J9CodeGenerator.cpp#L5371-L5372)